src/HttpServer.gren

```diff
@@ -0,0 +1,288 @@
+    { urlProtocol : String
+    , urlHost : String
+    , urlPort : String
+    , urlPath : String
+    , urlQuery : String
+    , urlFragment : String
+    , headers : Array String
+    , method : String
+    , body : String
+    }
+    -> Request
+toRequest 
+    { urlProtocol
+    , urlHost
+    , urlPort
+    , urlPath
+    , urlQuery
+    , urlFragment
+    , headers
+    , method
+    , body 
+    } =
+    { method = toMethod method
+    , body = body
+    , headers =
+        headers
+            |> arrayPairs
+            |> dictFromPairs
+    , url =
+        { protocol = 
+            if urlProtocol == "https:" then
+                Https
+            else
+                Http
+        , host = urlHost
+        , path = urlPath
+        , port_ = String.toInt urlPort
+        , query = 
+            if urlQuery == "" then
+                Nothing
+            else
+                Just urlQuery
+        , fragment =
+            if urlFragment == "" then
+                Nothing
+            else
+                Just urlFragment
+        }
+    }
+
+
+toMethod : String -> Method
+toMethod s =
+    case s of
+        "GET" ->
+            GET
+        
+        "HEAD" ->
+            HEAD
+
+        "POST" ->
+            POST
+        
+        "PUT" ->
+            PUT
+        
+        "DELETE" ->
+            DELETE
+        
+        "CONNECT" ->
+            CONNECT
+        
+        "TRACE" ->
+            TRACE
+        
+        "PATCH" ->
+            PATCH
+        
+        _ ->
+            UNKNOWN
+
+
+arrayPairs : Array String -> Array (Array String)
+arrayPairs a =
+    let
+        pair = 
+            Array.takeFirst 2 a
+        
+        rest =
+            Array.dropFirst 2 a
+        
+        allPairs =
+            [ pair ] ++ case rest of
+                [] ->
+                    []
+                
+                _ ->
+                    arrayPairs rest
+    in
+    allPairs
+
+
+dictFromPairs : Array (Array String) -> Dict String String
+dictFromPairs pairs =
+    let
+        mapper p =
+            case p of
+                [a, b] ->
+                    Just { key = a, value = b }
+                
+                _ ->
+                    Nothing
+        
+        array =
+            Array.filterMap mapper pairs
+    in
+    Dict.fromArray array
+
+
+-- EFFECT STUFF
+
+
+type MySub msg
+    = OnRequestSub Server (Request -> Response -> msg)
+
+
+subMap : (a -> b) -> MySub a -> MySub b
+subMap f sub =
+    case sub of
+        OnRequestSub server msg ->
+            -- TODO: not sure this is right
+            OnRequestSub server (\req res -> f (msg req res))
+
+
+type alias State msg =
+    Array (MySub msg)
+
+
+init : Task Never (State msg)
+init =
+    Task.succeed []
+
+
+onEffects : Platform.Router msg event -> Array (MySub msg) -> State msg -> Task Never (State msg)
+onEffects router subs state =
+    -- TODO: not sure "remove all, then add" is the right or best strategy here...
```

Comment:
Yeah, It's tricky. You could try to decipher what I did for `src/Terminal.gren`.

If I remember correctly, I essentially wrap listeners in a `Process`, this allows me to use `Process.kill` when I detect that a sub is removed, which again calls a callback in kernel code that detaches the listener.

Of course, that also requires some book-keeping to be able to know _which_ process to kill 🤷

Reply:
Do you see any issue with the way I'm doing it here other than it not being efficient?

Reply:
Since you're doing it all at once (meaning it will happen within the same "tick" in JS's event loop) I think you should be fine.